source/extensions/tracers/skywalking/trace_segment_reporter.cc

```diff
@@ -0,0 +1,165 @@
+#include "extensions/tracers/skywalking/trace_segment_reporter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace SkyWalking {
+
+namespace {
+static constexpr uint32_t DEFAULT_DELAYED_SEGMENTS_CACHE_SIZE = 1024;
+
+// Convert SegmentContext to SegmentObject.
+TraceSegmentPtr toSegmentObject(const SegmentContext& segment_context) {
+  auto new_segment = std::make_unique<SegmentObject>();
+  SegmentObject& segment_object = *new_segment;
+
+  segment_object.set_traceid(segment_context.traceId());
+  segment_object.set_tracesegmentid(segment_context.traceSegmentId());
+  segment_object.set_service(segment_context.service());
+  segment_object.set_serviceinstance(segment_context.serviceInstance());
+
+  for (const auto& span_store : segment_context.spanList()) {
+    if (!span_store->sampled()) {
+      continue;
+    }
+    auto* span = segment_object.mutable_spans()->Add();
+
+    span->set_spanlayer(SpanLayer::Http);
+    span->set_spantype(span_store->isEntrySpan() ? SpanType::Entry : SpanType::Exit);
+    span->set_componentid(6000);
+
+    if (!span_store->peerAddress().empty()) {
+      span->set_peer(span_store->peerAddress());
+    }
+
+    span->set_spanid(span_store->spanId());
+    span->set_parentspanid(span_store->parentSpanId());
+
+    span->set_starttime(span_store->startTime());
+    span->set_endtime(span_store->endTime());
+
+    span->set_iserror(span_store->isError());
+
+    span->set_operationname(span_store->operation());
+
+    auto& tags = *span->mutable_tags();
+    tags.Reserve(span_store->tags().size());
+
+    for (auto& span_tag : span_store->tags()) {
+      tags.Add(std::move(const_cast<Tag&>(span_tag)));
+    }
+
+    SpanContext* previous_span_context = segment_context.previousSpanContext();
+    if (!previous_span_context) {
+      continue;
+    }
+
+    auto* ref = span->mutable_refs()->Add();
+    ref->set_traceid(previous_span_context->trace_id_);
+    ref->set_parenttracesegmentid(previous_span_context->trace_segment_id_);
+    ref->set_parentspanid(previous_span_context->span_id_);
+    ref->set_parentservice(previous_span_context->service_);
+    ref->set_parentserviceinstance(previous_span_context->service_instance_);
+    ref->set_parentendpoint(previous_span_context->endpoint_);
+    ref->set_networkaddressusedatpeer(previous_span_context->target_address_);
+  }
+  return new_segment;
+}
+
+const Http::LowerCaseString& authenticationTokenKey() {
+  CONSTRUCT_ON_FIRST_USE(Http::LowerCaseString, "Authentication");
+}
```

Comment:
nit. Can we put this at the top? Probably after the `DEFAULT_DELAYED_SEGMENTS_CACHE_SIZE`.